File: src/BrewUp.Rest.FSharp/Modules/PurchaseModule.fs
namespace BrewUp.Rest.FSharp.Modules

open Microsoft.AspNetCore.Builder
open Microsoft.Extensions.DependencyInjection
open Giraffe
open BrewUp.Purchase.Facade.FSharp

type PurchaseModule() =
    interface IModule with
        member _.IsEnabled = true
        member _.Order = 0
        
        member _.Register(builder: WebApplicationBuilder) =
            builder.Services.AddScoped<IPurchaseFacade, PurchaseFacadeHelper>() |> ignore
            builder.Services
            
        member _.Configure(app: WebApplication) =
            let purchaseModule = PurchaseEndpoints.PurchaseModule()
            let routes = (purchaseModule :> IModule).RegisterRoutes()
            app.UseGiraffe(routes)
            app

File: src/BrewUp.Rest/Modules/PurchaseModule.cs
using BrewUp.Purchase.Domain;
using BrewUp.Purchase.Facade;
using BrewUp.Purchase.Infrastructure;

namespace BrewUp.Rest.Modules;

public static class PurchaseModule
{
  public static IServiceCollection Register(WebApplicationBuilder builder)
  {
    builder.Services.AddScoped<IPurchaseFacade, PurchaseFacade>();

    builder.Services.AddPurchaseDomain();
    builder.Services.AddPurchaseInfrastructure();

    return builder.Services;
  }

  public static WebApplication Configure(WebApplication app)
  {
    var group = app.MapGroup("/v1/purchase")
      .WithTags("Purchase");

    group.MapGet("/", () => Results.Ok("Purchase module is running"))
      .WithName("GetPurchaseStatus")
      .WithSummary("Get Purchase module status")
      .WithDescription("Returns the status of the Purchase module");

    return app;
  }
}